fileglob: make fileglob work with globbed path components again (Fixes: #17136)

rupran opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lookup/fileglob plugin

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (devel a10209cbe9) last updated 2016/12/27 22:05:26 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

PR #16630 broke globbing in with_fileglob for globbed path components
(that is, if an asterisk does not appear in the file name but in one of the 
path components higher up) by changing from path_dwim_relative to
path_dwim_relative_stack. While the former does something like this (for
a path specification like "/tmp/*/some.conf", see Issue #17136 for an
example playbook - this playbook is also shown below)

```python
search = ["/tmp/*"]
for candidate in search:
    if os.path.exists(candidate):
        break
return candidate
```

the check inside path_dwim_relative_stack works slightly differently:

```python
result = None
for candidate in search:
    if os.path.exists("/tmp/*"):
        result = "/tmp/*"
        break
return result
```

The way the loop in the former example works makes path_dwim_relative
return the last element from the 'search' list even if os.path.exists
never returned True, while path_dwim_relative_stack returns None in
the same case. This ultimately makes the fileglob plugin fail because
find_file_in_search_path returns None now.

To allow asterisks in paths again, this change breaks the search for
potential targets out of path_dwim_relative_stack. This function can
then be called from the fileglob plugin to find all potential target
directories without checking if these dirs (like "/tmp/*") actually
exist. fileglob will now apply globbing first and check for the actual
existence of the files _after_ globbing.

- Example playbook (from Issue #17136):
```yaml
---
- name: list files
  hosts: localhost
  tasks:
    - name: create test dirs
      file: state=directory path={{ item }}
      with_items:
        - /tmp/foo
        - /tmp/bar

    - name: create test files
      file: state=touch path={{ item }}
      with_items:
        - /tmp/foo/some.conf
        - /tmp/bar/some.conf

    - debug: var=item
      with_fileglob: "/tmp/*/some.conf"
```

- Output without this patch (in 2.2.0):
```
TASK [debug] *******************************************************************
 [WARNING]: Unable to find '/tmp/*' in expected paths.
```

- Output before 2.2.0 (more specifically: before MR #16630) and with this patch:
```
TASK [debug] *******************************************************************
ok: [localhost] => (item=/tmp/foo/some.conf) => {
    "item": "/tmp/foo/some.conf"
}
ok: [localhost] => (item=/tmp/bar/some.conf) => {
    "item": "/tmp/bar/some.conf"
}
```

rupran commented:
What does the failing check have something to do with this change?

Only one test fails, and this shows an error like ''Error validating the server's certificate for https://gist.githubusercontent.com/abadger/...". Can I somehow trigger a retest?

rupran committed:
fileglob: make fileglob work with globbed path components again

PR #16630 broke globbing in with_fileglob for globbed path components
by changing from path_dwim_relative to path_dwim_relative_stack. While
the former does something like this (for a path specification like
"/tmp/*/some.conf", see Issue #17136 for an example playbook)

search = ["/tmp/*"]
for candidate in search:
    if os.path.exists(candidate):
        break
return candidate

the check inside path_dwim_relative_stack works slightly differently:

result = None
if os.path.exists("/tmp/*"):
    result = "/tmp/*"
return result

The way the loop in the former example works makes path_dwim_relative
return the last element from the 'search' list even if os.path.exists
never returned True, while path_dwim_relative_stack returns None in
the same case. This ultimately makes the fileglob plugin fail because
find_file_in_search_path returns None now.

To allow asterisks in paths again, this change breaks the search for
potential targets out of path_dwim_relative_stack. This function can
then be called from the fileglob plugin to find all potential target
directories without checking if these dirs (like "/tmp/*") actually
exist. fileglob will now apply globbing first and check for the actual
existence of the files _after_ globbing.

Additionally, fileglob now calls os.path.normpath() on the results of
glob.glob and eliminates duplicate results.

luto commented:
This fixes the issue in `stable-2.2` and `devel` for me. :+1:

gundalow commented:
> [19:09] < sivel> btw, that PR may also help with https://github.com/ansible/ansible/issues/18549

alikins commented:
Patch looks reasonable to me, but haven't looked close yet. I'm in favor of any changes that make the dataloader path handling easier to digest.

alikins commented on lib/ansible/plugins/lookup/fileglob.py:
```diff
@@ -32,8 +33,16 @@ def run(self, terms, variables=None, **kwargs):
+            term_dir = os.path.dirname(term)
+            if term_dir.startswith('~') or term_dir.startswith(os.path.sep):
+                basepath = unfrackpath(to_text(term_dir))
+                candidates = [basepath]
```
Could this ~ and / handling be moved into path_dwim_get_candidates?

rupran replied:
I guess I could do the same thing as above (building a one-item list as candidates for paths starting in ~ or /) in path_dwim_get_candidates, which would then leave path_dwim_relative_stack as merely a) getting the candidates and b) the check for existence of a candidate in the result of a).

Should I upload that change as an amended commit, or should we wait for more comments?

rupran replied:
I added your suggestion as a separate commit for now, thanks for your input!

mattclay commented:
Closing and re-opening to trigger CI.

rupran committed:
fileglob: move checking for ~ and / into path_dwim_get_candidates

Instead of checking for paths relative to the home directory or
absolute paths twice (in path_dwim_relative_stack and in the fileglob
plugin), move the whole check into path_dwim_get_candidates and use
only this function in the fileglob plugin.

This change only looks this huge in dataloader.py because the
indentation changed and the  identical comments make git lose the
correct context for the code. Actually, this only moves the check
for ~ and / around and changes the parameter because b_source is not
needed in path_dwim_relative_stack anymore.

rupran commented:
bot_status

gundalow commented:
Rebase needed

luto commented:
I hit a snag with this. Consider the following:

```
.
├── roles
│   ├── child
│   │   └── tasks
│   │       └── main.yml
│   └── main
│       └── meta
│           └── main.yml
└── site.yml
```

`site.yml`:
```yml
- hosts: localhost
  roles: ['main']
```

`roles/child/tasks/main.yml`:
```yml
- debug: var=item
  with_fileglob: '../child/tasks/*.yml'
```

`roles/main/meta/main.yml`:
```yml
dependencies:
  - child
```

Ansible 2.1 generates the following:
```
ok: [localhost] => (item=/home/luto/uberspace/ansible/foo/roles/rr/tasks/main.yml) => {
    "item": "/home/luto/uberspace/ansible/foo/roles/rr/tasks/main.yml"
}
```
It does not consider paths relative to the `main`-role.

Ansible `devel` with this patch generates this:
```
ok: [localhost] => (item=/home/luto/uberspace/ansible/foo/roles/child/../child/tasks/main.yml) => {
    "item": "/home/luto/uberspace/ansible/foo/roles/child/../child/tasks/main.yml"
}
ok: [localhost] => (item=/home/luto/uberspace/ansible/foo/roles/main/../child/tasks/main.yml) => {
    "item": "/home/luto/uberspace/ansible/foo/roles/main/../child/tasks/main.yml"
}
```
It *does* consider paths relative to the `main`-role and thus yields two results. That's different to the previous behavior, but fine by me. The problem is that those two paths are the actually the same. So we've got duplicate results now. :-/

rupran commented:
@luto: The difference between the two is basically the "circumvention" of this `break` statement: 

https://github.com/ansible/ansible/pull/19297/commits/5cbdac07b62341881413679786fce13481b7cd74#diff-b445cdc457c1d402b8d059e9f89b5a0fL357

`path_dwim_relative_stack` stops the iteration over the possible candidate directories as soon as it finds the first existing result, and the child role seems to be the first one it hits.

Printing the contents of `search` just before the loop gives me the following paths (highlighted with `<=` is the one where the iteration stops in Ansible 2.1):

```
/tmp/ansible-test/roles/child/files/../child/tasks
/tmp/ansible-test/roles/child/../child/tasks                       <=
/tmp/ansible-test/roles/main/files/../child/tasks
/tmp/ansible-test/roles/main/../child/tasks
/tmp/ansible-test/roles/child/tasks/files/../child/tasks
/tmp/ansible-test/roles/child/tasks/../child/tasks
/tmp/ansible-test/files/../child/tasks
/tmp/ansible-test/../child/tasks
```

Recreating that behaviour would be a matter of breaking out of the loop in the fileglob plugin after the first success of `glob.glob` instead of breaking in `path_dwim_relative_stack`, i.e. doing something like this:

```python
globbed = glob.glob(to_bytes(os.path.join(dwimmed_path, term_file), errors='surrogate_or_strict'))
if globbed:
    ret.extend(to_text(g, errors='surrogate_or_strict') for g in globbed if os.path.isfile(g))
    break
```

gundalow commented:
Is this needed for 2.3?

luto commented:
@rupran maybe the list of paths could be normalized and then be made unique, after all possible paths have been collected?

luto commented:
@gundalow that would be nice.. what's the deadline for that?

rupran commented:
@luto Sure, we could use a `set` in the plugin to aggregate every result and call `os.path.realpath` on every result of `glob.glob` before adding it... I will code that up later.

gundalow commented:
@luto 2.3 RC3 will be tomorrow (Thursday 30th), unsure if their will be an RC4...

luto commented:
@gundalow works fine for me now with the recent changes. Just needs a review from your side.

```
TASK [child : debug] ************************************************************************************
ok: [localhost] => (item=/home/luto/uberspace/ansible/foo/roles/child/tasks/main.yml) => {
    "item": "/home/luto/uberspace/ansible/foo/roles/child/tasks/main.yml"
}
```
@rupran thanks!

gundalow commented:
Alternative fix proposed in https://github.com/ansible/ansible/pull/23265

bcoca commented:
closing in favor of the `filetree` plugin
